src/model_theory/basic.lean

```diff
@@ -0,0 +1,342 @@
+  to the `L`-structure `N` that commutes with the interpretations of functions, and which preserves
+  the interpretations of relations in both directions.
+* A `first_order.language.equiv`, denoted `M ≃[L] N`, is an equivalence from the `L`-structure `M`
+  to the `L`-structure `N` that commutes with the interpretations of functions, and which preserves
+  the interpretations of relations in both directions.
+
+## References
+For the Flypitch project:
+- [J. Han, F. van Doorn, *A formal proof of the independence of the continuum hypothesis*]
+[flypitch_cpp]
+- [J. Han, F. van Doorn, *A formalization of forcing and the unprovability of
+the continuum hypothesis*][flypitch_itp]
+
+-/
+
+universes l u
+
+variables {α : Type u} {β : Type u}
+
+namespace first_order
+
+/-- A first-order language consists of a type of functions of every natural-number arity and a
+  type of relations of every natural-number arity. -/
+structure language : Type (l+1) :=
+(functions : ℕ → Type l) (relations : ℕ → Type l)
+
+namespace language
+
+/-- The empty language has no symbols. -/
+def empty : language := ⟨λ _, pempty, λ _, pempty⟩
+
+instance : inhabited language := ⟨empty⟩
+
+/-- The type of constants in a given language. -/
+@[nolint has_inhabited_instance] def const (L : language.{l}) : Type l := L.functions 0
+
+variable (L : language.{u})
+
+/-- A language is relational when it has no function symbols. -/
+class is_relational : Prop :=
+(empty_functions : ∀ n, L.functions n → false)
+
+/-- A language is algebraic when it has no relation symbols. -/
+class is_algebraic : Prop :=
+(empty_relations : ∀ n, L.relations n → false)
+
+variable {L}
+
+lemma is_relational_of_empty {symb : ℕ → Type u}  : is_relational ⟨λ _, pempty, symb⟩ :=
+⟨by { intro n, apply pempty.elim }⟩
+
+lemma is_algebraic_of_empty {symb : ℕ → Type u}  : is_algebraic ⟨symb, λ _, pempty⟩ :=
+⟨by { intro n, apply pempty.elim }⟩
+
+variables (L) (M : Type u)
+
+/-- A first-order structure on a type `M` consists of interpretations of all the symbols in a given
+  language. Each function of arity `n` is interpreted as a function sending tuples of length `n`
+  (modeled as `(fin n → M)`) to `M`, and a relation of arity `n` is a function from tuples of length
+  `n` to `Prop`. -/
+class Structure :=
+(fun_map : ∀{n}, L.functions n → (fin n → M) → M)
+(rel_map : ∀{n}, L.relations n → (fin n → M) → Prop)
+
+variables (N : Type u) [L.Structure M] [L.Structure N]
+
+open first_order.language.Structure
+
+/-- A homomorphism between first-order structures is a function that commutes with the
+  interpretations of functions and maps tuples in one structure where a given relation is true to
+  tuples in the second structure where that relation is still true. -/
+protected structure hom :=
+(to_fun : M → N)
+(map_fun' : ∀{n} (f : L.functions n) x, to_fun (fun_map f x) = fun_map f (to_fun ∘ x))
+(map_rel' : ∀{n} (r : L.relations n) x, rel_map r x → rel_map r (to_fun ∘ x))
```

Comment:
Perhaps it makes sense to add some auto_param's to these fields?
Same goes for the fields in embedding?

Reply:
All I know how to do in that world is to add ` . obviously`, and I'm not sure if it'll work well, but I think it's worth a try.
In the future, I will write some more connective tissue to make these easy to prove, such as specific code for relational or algebraic situations, and situations with only constants, unary, and binary symbols.

Reply:
Adding `. obviously` seems pretty great to me. I went and deleted the things that that took care of - it reduced the line count, but might make things slower. If people want, I can add them back in.

Reply:
I'm a fan. :-)